src/PackageExtensionCompat.jl

```diff
@@ -5,39 +5,55 @@ export @require_extensions
 
 @static if !HAS_NATIVE_EXTENSIONS
-    using Requires, TOML
+    using MacroTools, Requires, TOML
 
-    function rewrite_import(str, pkgs)
-        parts = split(strip(str))
-        if length(parts) == 1 || (length(parts) ≥ 2 && parts[2] == "as")
-            if parts[1] ∈ pkgs
-                parts[1] = string("..", parts[1])
+    @static if hasmethod(Base.include, Tuple{Function, Module, String})
+        function _include(mapexpr::Function, m::Module, path::AbstractString)
+            Base.include(mapexpr, m, path)
+        end
+    else
+        function _include(mapexpr::Function, m::Module, path::AbstractString)
+            path = abspath(path)
+            cd(dirname(path)) do
+                str = read(basename(path), String)
+                pos = 1
+                while true
+                    (expr, pos) = Meta.parse(str, pos; raise=false)
+                    expr !== nothing || break
+                    m.eval(mapexpr(expr))
+                end
             end
         end
-        join(parts, " ")
     end
 
-    function rewrite_imports(str, pkgs)
-        parts = split(str, ",")
-        parts = map(part -> rewrite_import(part, pkgs), parts)
-        join(parts, ", ")
+    function rewrite(toplevel::Module, pkgs)
+        Base.Fix1(MacroTools.postwalk, block -> rewrite_block(block, toplevel, pkgs))
     end
 
-    function rewrite_line(line, pkgs)
-        pat = r"^(\s*(using|import)\s+)([^;:#$]*[^;:#$\s])(.*)$"
-        m = match(pat, line)
-        if m === nothing
-            line
+    function rewrite_block(block, toplevel::Module, pkgs)
+        if Meta.isexpr(block, :call) && block.args[1] == :include
+            # inner include, rewrite it recursively
+            local_mod = Expr(:macrocall, Symbol("@__MODULE__"), @__LINE__)
+            Expr(:call, _include, rewrite(toplevel, pkgs), local_mod, block.args[2])
+        elseif Meta.isexpr(block, [:using, :import])
+            # using or import block, replace references to pkgs
+            imports = map(block.args) do use
+                Meta.isexpr(use, [:(:), :as]) ?
+                    Expr(use.head,
+                         rewrite_use(use.args[1], toplevel, pkgs),
+                         use.args[2:end]...) :
+                    rewrite_use(use, toplevel, pkgs)
+            end
+            Expr(block.head, imports...)
         else
-            string(m.captures[1], rewrite_imports(m.captures[3], pkgs), m.captures[4])
+            # leave everything else alone
+            block
         end
     end
 
-    function rewrite(srcfile, trgfile, pkgs)
-        lines = readlines(srcfile)
-        lines = map(line -> rewrite_line(line, pkgs), lines)
-        code = join(lines, "\n")
-        write(trgfile, code)
+    function rewrite_use(use::Expr, toplevel::Module, pkgs)::Expr
+        @assert Meta.isexpr(use, :.)
+        string(use.args[1]) ∈ pkgs ? Expr(:., nameof(toplevel), use.args...) : use
```

Comment:
This rewrites `using WeakDep` to `using TopPkg.WeakDep`? Can you leave a comment to that effect?

I was a bit worried this might not be as robust as `using ..WeakDep` (i.e. what if `using TopPkg` doesn't work) but I think it's actually fine because the `TopPkg` *must* be an installed package in order to use package extensions, and therefore `using TopPkg` will always succeed. Is this right? Any cases where this wouldn't work?

Reply:
Yes, you figured that out correctly. I'll add a comment.

After giving this some thought, I now think that `TopPkg.WeakDep` is a safer way of referring to the dependency than `..WeakDep`, because it works from any level of the module hierarchy.

I can think of a somewhat forced scenario where this possibly wouldn't work:

```
module TopPkg
   module SubPkg
       const TopPkg = nothing
       # wouldn't work here since TopPkg doesn't refer to the package locally
       using WeakDep
   end
end
```

But I wouldn't expect anyone in their right mind to hide `TopPkg` like that.

Other than that, `TopPkg` should be reachable using the package manager, AFAIK.

Reply:
Do you mean
```julia
const TopPkg = nothing
using TopPkg.WeakDep
```
?

I think even this will work. AFAIU `using X` always imports the package `X` regardless of the local scope. It's only things like `using .X` which depend on the scope. So in conclusion I think we're good.

Reply:
```
using TopPkg.WeakDep
```

That's what it would be rewritten to, correct. And you are probably right about `using` ignoring the local scope.